added deprecate_nonkeyword_arguments to function where

Jiezheng2018 opened the pull request:
- xref  #41486
- generic.where. tests added / passed
- Ensure all linting tests pass, see [here](https://pandas.pydata.org/pandas-docs/dev/development/contributing.html#code-standards) for how to run them.  
- whatsnew entry. -- added.

Commit:
added deprecate_nonkeyword_arguments to function where

MarcoGorelli commented on pandas/core/generic.py:
```diff
@@ -9077,6 +9078,7 @@ def _where(
         name="where",
         name_other="mask",
     )
+    @deprecate_nonkeyword_arguments(version="2.0", allowed_args=["self", "labels"])
```
let's allow self, cond, and other here

Jiezheng2018 replied:
I made some corrections and did 'git push origin shiny-new-feature', but I can't see the pull request is updated. What have I missed?

Jiezheng2018 replied:
I also changed the above to be like this:
`@deprecate_nonkeyword_arguments(version="2.0", allowed_args=["self", "cond", "other"])`
Is that what you mean? I couldn't find a way to quote the code as you did, so I just copied the code here.

MarcoGorelli replied:
yup, that's what I meant :+1:

Jiezheng2018 replied:
Thanks a lot. I deleted my previous comments as I realized I had a problem with my local git. Now I have added what's new, and the test. Still not sure I understood this issue fully.

MarcoGorelli replied:
> Still not sure I understood this issue fully.

Try running:
```python
In [1]: ser = pd.Series([1,2,3])

In [2]: ser.where(ser>1, 0, False)
Out[2]: 
0    0
1    2
2    3
dtype: int64
```

It works fine - however, in a future release, we'd like to force users to pass arguments (apart from `cond` and `other`) by name, i.e. `ser.where(ser>1, 0, inplace=True)`. But before we do that, we need a FutureWarning. The decorator you're adding here should change the above so that this will happen:
```python
In [1]: ser = pd.Series([1,2,3])

In [2]: ser.where(ser>1, 0, False)
<ipython-input-2-7bc00a67a7aa>:1: FutureWarning: In a future version of pandas all arguments of NDFrame.where except for the arguments 'cond' and 'other' will be keyword-only
  ser.where(ser>1, 0, False)
Out[2]: 
0    0
1    2
2    3
dtype: int64
```

Make sure to fetch and merge latest changes for this to work properly though:
```
git fetch upstream
git merge upstream/master
```

Jiezheng2018 replied:
Thanks a lot for the detailed explanation. I think I understand it now. 
My test should be: 
' s = Series(range(5))
  with tm.assert_produces_warning(FutureWarning):
         s.where(s > 1, 10, False)'
because 'False' was passed in as a non-keyword (my understand of non-keyword is the same as a positional argument, is this correct?), the future warning will be triggered. 
I hope I got it right this time

MarcoGorelli reviewed (changes requested):
Thanks @Jiezheng2018 , this is off to a good start - if you look at https://github.com/pandas-dev/pandas/pull/41486 you'll see an example of how to add a whatsnew note and a test

Commit:
added test

Jiezheng2018 committed:
Merge branch 'pandas-dev:master' into shiny-new-feature

Commit:
updated generic.py where function

Commit:
added line 651 in whatsnew v1.3.0.rst

MarcoGorelli commented on pandas/tests/generic/test_generic.py:
```diff
@@ -339,6 +339,11 @@ def test_copy_and_deepcopy(self, shape, func):
         assert obj_copy is not obj
         self._compare(obj_copy, obj)
 
+    def test_where(self, cond):
```
you don't need `cond` as an argument here - you can try running this test with
```
pytest pandas/tests/generic/test_generic.py
```

Jiezheng2018 replied:
I run into a lot of problems when I tried running pytest. 

1. Using terminal I get all green test passed:
'
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % pytest pandas/tests/generic/test_generic.py
=========================================================== test session starts ===========================================================
platform darwin -- Python 3.8.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /Users/jiezheng/pandas-jie, configfile: pyproject.toml
plugins: hypothesis-6.13.0, cov-2.12.0, xdist-2.2.1, asyncio-0.14.0, instafail-0.4.1, forked-1.3.0
collected 22 items                                                                                                                        

pandas/tests/generic/test_generic.py ......................

-------------------------------------- generated xml file: /Users/jiezheng/pandas-jie/test-data.xml ---------------------------------------
========================================================== slowest 30 durations ===========================================================
0.07s call     pandas/tests/generic/test_generic.py::TestNDFrame::test_depr_take_kwarg_is_copy[DataFrame-True]
0.01s call     pandas/tests/generic/test_generic.py::TestNDFrame::test_squeeze

(28 durations < 0.005s hidden.  Use -vv to show these durations.)
=========================================================== 22 passed in 0.17s ============================================================
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % 
' 
2. running on terminal just on my function, then I get the not found error. 

' (pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % pytest pandas/tests/generic/test_generic.py::test_where
=========================================================== test session starts ===========================================================
platform darwin -- Python 3.8.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /Users/jiezheng/pandas-jie, configfile: pyproject.toml
plugins: hypothesis-6.13.0, cov-2.12.0, xdist-2.2.1, asyncio-0.14.0, instafail-0.4.1, forked-1.3.0
collected 0 items                                                                                                                         

-------------------------------------- generated xml file: /Users/jiezheng/pandas-jie/test-data.xml ---------------------------------------
========================================================== no tests ran in 0.01s ==========================================================
ERROR: not found: /Users/jiezheng/pandas-jie/pandas/tests/generic/test_generic.py::test_where
(no name '/Users/jiezheng/pandas-jie/pandas/tests/generic/test_generic.py::test_where' in any of [<Module test_generic.py>])

(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % 

3. when using vscode, I get the following error message : 

`
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % pytest pandas/tests/generic/test_generic.py
ImportError while loading conftest '/Users/jiezheng/pandas-jie/pandas/conftest.py'.
pandas/__init__.py:22: in <module>
    from pandas.compat import (
pandas/compat/__init__.py:15: in <module>
    from pandas.compat.numpy import (
pandas/compat/numpy/__init__.py:7: in <module>
    from pandas.util.version import Version
pandas/util/__init__.py:1: in <module>
    from pandas.util._decorators import (  # noqa
pandas/util/_decorators.py:14: in <module>
    from pandas._libs.properties import cache_readonly  # noqa
pandas/_libs/__init__.py:13: in <module>
    from pandas._libs.interval import Interval
E   ModuleNotFoundError: No module named 'pandas._libs.interval'
`
Any of these are normal? I will take the 'cond' out by the way.

MarcoGorelli replied:
Nice! Should be good then. For 2) you probably need to specify the class name as well, or you could try `pytest pandas/tests/generic/test_generic.py -k test_where` . For 3) I'm surprised that it worked in the terminal but not in vscode, but if you see an error like that it typically means you need to build the c extensions again, `python setup.py build_ext -i -j 4`

Jiezheng2018 replied:
for 2) this is what I get now: 
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % pytest pandas/tests/generic/test_generic.py -k test_where
============================= test session starts ==============================
platform darwin -- Python 3.8.10, pytest-6.2.4, py-1.10.0, pluggy-0.13.1
rootdir: /Users/jiezheng/pandas-jie, configfile: pyproject.toml
plugins: hypothesis-6.13.0, cov-2.12.0, xdist-2.2.1, asyncio-0.14.0, instafail-0.4.1, forked-1.3.0
collected 22 items / 22 deselected                                             

--------- generated xml file: /Users/jiezheng/pandas-jie/test-data.xml ---------
============================ 22 deselected in 0.03s ============================
I thought I should get something like 1 test passed?  Not sure how to interpret the result here? 

for 3) I tried in vscode to build the c extension using both base and panda-env 

> 

(base) jiezheng@Jies-MacBook-Air pandas-jie % python setup.py build_ext -i -j 4
  File "setup.py", line 249
    f"{extension}-source file '{sourcefile}' not found.\n"
                                                         ^
SyntaxError: invalid syntax
(base) jiezheng@Jies-MacBook-Air pandas-jie % conda activate pandas-dev                              
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % python setup.py build_ext -i -j 4
  File "setup.py", line 249
    f"{extension}-source file '{sourcefile}' not found.\n"
                                                         ^
SyntaxError: invalid syntax

> 


`
and the line 249 refers to this run function. 

> 

    def run(self):
        if "cython" in cmdclass:
            self.run_command("cython")
        else:
            # If we are not running cython then
            # compile the extensions correctly
            pyx_files = [(self._pyxfiles, "c"), (self._cpp_pyxfiles, "cpp")]

            for pyxfiles, extension in pyx_files:
                for pyxfile in pyxfiles:
                    sourcefile = pyxfile[:-3] + extension
                    msg = (
                        f"{extension}-source file '{sourcefile}' not found.\n"
                        "Run 'setup.py cython' before sdist."
                    )
                    assert os.path.isfile(sourcefile), msg
        sdist_class.run(self)

> 

I then tried to run this:  python setup.py cython. -- still error 

> 

(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % python setup.py cython
  File "setup.py", line 249
    f"{extension}-source file '{sourcefile}' not found.\n"
                                                         ^
SyntaxError: invalid syntax

> 

Finally, I searched the whole folder and can't find anything called pyxfiles.  

At a total loss here and not sure what to do. I am using a mac with M1 chip. not sure whether would explain anything or not.

MarcoGorelli replied:
> I tried in vscode to build the c extension using both base and panda-env

No need to use `base`, just `pandas-dev`. Can you show your complete error message please, and also show the output of `git status` and `git diff origin/shiny-new-feature` to check you haven't accidentally changed something locally?

Jiezheng2018 replied:
Here is the whole error message: 

(base) jiezheng@Jies-MacBook-Air pandas-jie % conda activate pandas-dev                                
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % git status
On branch shiny-new-feature
Your branch is up to date with 'origin/shiny-new-feature'.

nothing to commit, working tree clean
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % git diff origin/shiny-new-feature
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % python setup.py build_ext -i -j 4
  File "setup.py", line 249
    f"{extension}-source file '{sourcefile}' not found.\n"
                                                         ^
SyntaxError: invalid syntax
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % pytest pandas/tests/generic/test_generic.py 
ImportError while loading conftest '/Users/jiezheng/pandas-jie/pandas/conftest.py'.
pandas/__init__.py:22: in <module>
    from pandas.compat import (
pandas/compat/__init__.py:15: in <module>
    from pandas.compat.numpy import (
pandas/compat/numpy/__init__.py:7: in <module>
    from pandas.util.version import Version
pandas/util/__init__.py:1: in <module>
    from pandas.util._decorators import (  # noqa
pandas/util/_decorators.py:14: in <module>
    from pandas._libs.properties import cache_readonly  # noqa
pandas/_libs/__init__.py:13: in <module>
    from pandas._libs.interval import Interval
E   ModuleNotFoundError: No module named 'pandas._libs.interval'
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie %

MarcoGorelli replied:
Thanks @Jiezheng2018 !

Just to confirm, have you already tried:
```
python setup.py build_ext --inplace --force
```
and
```
xcode-select --install
```
?

If so, @WillAyd sorry to tag you again, but I don't know what to suggest here

Jiezheng2018 replied:
'(base) jiezheng@Jies-MacBook-Air pandas-jie % conda activate pandas-dev                                               
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % python setup.py build_ext --inplace --force

  File "setup.py", line 249
    f"{extension}-source file '{sourcefile}' not found.\n"
                                                         ^
SyntaxError: invalid syntax
(pandas-dev) jiezheng@Jies-MacBook-Air pandas-jie % xcode-select --install

xcode-select: error: command line tools are already installed, use "Software Update" to install updates
'
The above is what I get when running them in VS Code, but on the terminal, it will rebuild setup.py successfully. 
Maybe I need to reinstall my VS Code?

MarcoGorelli replied:
I don't know, but if it's working from the terminal let's just use that for now

MarcoGorelli reviewed (changes requested):
Getting there! :muscle:

Commit:
removed cond from test_where

Commit:
changed test to be s.where(s > 1, 10, False) to triger warning

Jiezheng2018 committed:
Merge branch 'master' into shiny-new-feature

MarcoGorelli commented on pandas/core/generic.py:
```diff
@@ -9082,6 +9083,9 @@ def _where(
         name_other="mask",
     )
+    @deprecate_nonkeyword_arguments(
+        version="2.0", allowed_args=["self", "cond", "other"]
```
Can you put `version=None` please? I know the original PR I opened had this, but then other reviewers preferred to not specify the version number

Jiezheng2018 replied:
`
    @deprecate_nonkeyword_arguments(
        version=None, allowed_args=["self", "cond", "other"]
    )
`.  Right?  I have updated my pull request for this now

Jiezheng2018 replied:
How did you reference the code with the line number in by the way? I tried insert quote but am not getting the same view as yours?

MarcoGorelli replied:
Looks right, will take a closer look after work!

Like this: https://youtu.be/5IDpsBccqLk

MarcoGorelli commented on pandas/tests/generic/test_generic.py:
```diff
@@ -339,6 +339,11 @@ def test_copy_and_deepcopy(self, shape, func):
         assert obj_copy is not obj
         self._compare(obj_copy, obj)
 
+    def test_where(self):
+        s = Series(range(5))
+        with tm.assert_produces_warning(FutureWarning):
+            s.where(s > 1, 10, False)
+
```
can you put the test in `pandas/tests/frame/indexing/test_where.py` and `pandas/tests/series/indexing/test_where.py`?

Jiezheng2018 replied:
will do this later today. Getting some flake8 errors and I will learn and fix it.

MarcoGorelli reviewed (changes requested):
Thanks @Jiezheng2018 for updating!

Just left a couple of comments - please have a look at https://github.com/pandas-dev/pandas/pull/41510 for the most up-to-date example of how to do this, as that PR got merged

Commit:
changed version to None

Commit:
deleted test_where from test_generic.py

Commit:
added test_where to frame folder

Commit:
added test_where to frame folder

Commit:
added test_where to series folder

Jiezheng2018 commented:
It has failed all Azure tests on py37_32bit. Not sure what I should do now?

MarcoGorelli commented on doc/source/whatsnew/v1.3.0.rst:
```diff
@@ -649,7 +649,7 @@ Deprecations
 - Deprecated using ``usecols`` with out of bounds indices for ``read_csv`` with ``engine="c"`` (:issue:`25623`)
 - Deprecated passing arguments as positional (except for ``"method"``) in :meth:`DataFrame.interpolate` and :meth:`Series.interpolate` (:issue:`41485`)
-
+- Deprecated non-keyword arguments to ``where`` in :class:`NDFrame` (:issue:`41523`)
```
mention `(other than ``"cond"`` and ``"other"`)`

MarcoGorelli replied:
Let's mention `(other than ``"cond"`` and ``"other"`)`

MarcoGorelli commented on pandas/tests/frame/indexing/test_where.py:
```diff
@@ -745,3 +745,11 @@ def test_where_bool_comparison():
     tm.assert_frame_equal(result, expected)
+
+
+def test_where():
+
+    s = Series(range(5))
+
+    with tm.assert_produces_warning(FutureWarning):
+        s.where(s > 1, 10, False)
```
A few comments here:

1. this is in pandas/tests/frame/index/test_where.py, but here we have a series - could you make this test use a DataFrame instead?
2. could you check the warning message that the test produces? See the PR I linked to for an example
3. Could you also check the result? So here it would be
    ```
    with tm.assert_produces_warning(FutureWarning):
        result = s.where(s > 1, 10, False)
    expected = Series([10, 2, 3])
    tm.assert_series_equal(result, expected)
   ```
4. Could you give the test a more descriptive title?
5. In the line under `test_where`, could you put a comment with the issue number of the issue this is addressing?

MarcoGorelli commented on pandas/tests/series/indexing/test_where.py:
```diff
@@ -140,6 +140,9 @@ def test_where():
     rs = s2.where(cond[:3], -s2)
     tm.assert_series_equal(rs, expected)
 
+    with tm.assert_produces_warning(FutureWarning):
+        s.where(s > 1, 10, False)
```
Let's make a new test case for this, otherwise existing ones get longer and longer

MarcoGorelli commented on pandas/core/generic.py:
```diff
@@ -9082,6 +9083,9 @@ def _where(
         name="where",
         name_other="mask",
     )
+    @deprecate_nonkeyword_arguments(
+        version=None, allowed_args=["self", "cond", "other"]
+    )
```
In order for the warning to show the correct message, `where` should be defined in `pandas/core/series.py` and `pandas/core/frame.py` as well, and the `deprecate_nonkeyword_arguments` set there as well - see https://github.com/pandas-dev/pandas/pull/41510 for an example of that

MarcoGorelli reviewed (changes requested):
Thanks @Jiezheng2018, just a few more comments!

The test failure is caused by the stacklevel of another warning inside `where` not being set correctly - pretty sure it'll be enough to just increase the stacklevel to 3 in the lines https://github.com/pandas-dev/pandas/blob/b3e335254f46a526ee3ce9bb757eac4011d9d1fe/pandas/core/generic.py#L9253-L9258

Jiezheng2018 commented:
Thanks a lot for the comments :) I will work on them over the weekend. 

(P.S. I am really learning a lot by reading up your comments and documents and fixing error code.  Really appreciated your patience with me. I might respond a bit slower in the coming week but be assured that I am commited to get this issue done :)

MarcoGorelli reviewed (changes requested):
Glad to hear you're learning a lot from this :smile: 

In a different PR (https://github.com/pandas-dev/pandas/pull/41580#issuecomment-847241393), Jeff has asked for `other` to be positional, so let's change this to be `allowed_args=["self", "cond"]` - sorry for having midled you!

----

EDIT: you'd already set it as positional, so my comment here was worse than useless 🤦  My bad for leaving a review when I hadn't slept that well...

Jiezheng2018 commented:
> Glad to hear you're learning a lot from this 😄
> 
> In a different PR ([#41580 (comment)](https://github.com/pandas-dev/pandas/pull/41580#issuecomment-847241393)), Jeff has asked for `other` to be positional, so let's change this to be `allowed_args=["self", "cond"]` - sorry for having midled you!

No problem at all. I haven't made much progress because I kept running into problems each time I do a git fetch upstream and git merge upstream/master. (My understanding is that I should do this each time before I start working, right?) 

I can't merge successfully because other people are also working on the same file (which is to be expected). However, what is really confusing for me is that the merge still fails even after I removed all my changes. Then it tells me that I have uncommitted changes. Trying git commit then I start failing on other pre-commit checks, either black, flake8 or absolufy-imports.    Am I doing something stupid here?

Commit:
resolve conflicts

Commit:
edited the whatsnew, mentioned other than cond and other

Commit:
changed stacklevel to 3 for futurewarning

Commit:
changed frame/index/test_where_non_keyword

Commit:
added where function in frame.py

Commit:
added where function in series.py

Commit:
changed allowed_args in frame

Commit:
changed allowed_args in series

Commit:
changed allowed_args in generic

Jiezheng2018 commented:
Hi, 
I think I have managed to complete all of your requests now. Do you mind having another look? I might miss or revert some of my previous work because of my git merge issues.

Commit:
adding github issue number

Jiezheng2018 commented:
Morning, it seems like I failed a lot of tests again. I had to change stacklevel back to 2, otherwise, I was failing all 'black' pre-commit tests with an error code of 123. I will do a bit more research and try to work on it later today or tomororw.

MarcoGorelli committed:
fixup

MarcoGorelli commented:
> Morning, it seems like I failed a lot of tests again. I had to change the future warning 'stack level' back to 2, otherwise, I was failing all 'black' pre-commit tests with an error code of 123. I will do a bit more research and try to work on it later today or tomorrow.

Hey @Jiezheng2018 - no worries, I've added a commit to fix some things up (partially to make sure this gets done in time, and partially because I'd misled you in some previous reviews 😳 ) - do you want to have a look at the recent changes?

> I had to change the future warning 'stack level' back to 2, otherwise, I was failing all 'black' pre-commit tests with an error code of 123

I'm pretty sure that's not the kind of thing the `black` formatter would complain about - perhaps you also changed something else and caused a syntax error by accident?

MarcoGorelli committed:
revert change to mask

MarcoGorelli committed:
fixup gh issue number

Jiezheng2018 commented:
> Glad to hear you're learning a lot from this 😄
> 
> In a different PR ([#41580 (comment)](https://github.com/pandas-dev/pandas/pull/41580#issuecomment-847241393)), Jeff has asked for `other` to be positional, so let's change this to be `allowed_args=["self", "cond"]` - sorry for having midled you!
> 
> EDIT: you'd already set it as positional, so my comment here was worse than useless 🤦 My bad for leaving a review when I hadn't slept that well...

No worries at all. :) Now I understand the positional keyword even better, was gonna read up on it more after work today. :)  thanks a lot

Jiezheng2018 commented:
> > Morning, it seems like I failed a lot of tests again. I had to change the future warning 'stack level' back to 2, otherwise, I was failing all 'black' pre-commit tests with an error code of 123. I will do a bit more research and try to work on it later today or tomorrow.
> 
> Hey @Jiezheng2018 - no worries, I've added a commit to fix some things up (partially to make sure this gets done in time, and partially because I'd misled you in some previous reviews 😳 ) - do you want to have a look at the recent changes?
> 
> > I had to change the future warning 'stack level' back to 2, otherwise, I was failing all 'black' pre-commit tests with an error code of 123
> 
> I'm pretty sure that's not the kind of thing the `black` formatter would complain about - perhaps you also changed something else and caused a syntax error by accident?

1) Thank you very much for the fixup. That's super helpful to see how you made the changes :) 
2) I am not very sure what I did that caused the 'black' to complain so much. I did notice sometimes exactly the same code on my VS code will fail the first time and pass the second time. I might need to re-install VS code at some point or just use the terminal instead.

MarcoGorelli reviewed (approved):
Looks good to me, will leave open for a bit in case others have comments

jreback commented:
thanks @Jiezheng2018
